File: src/python/messenger.py
import urllib2
import httplib
import json
import time

class Messenger:
  """Class for communicating with Paxos cluster"""
  def __init__(self, url):
    self.url = url

  def getNextMessage(self):
    while True:
      try:
        response = urllib2.urlopen(self.url)
        try:
          if response.code == 200:
            return json.load(response)
        finally:
          response.close()
      except ValueError as e:
        print e
        time.sleep(2.0)
      except urllib2.URLError as e:
        print e
        time.sleep(2.0)
      except httplib.BadStatusLine as e:
        print e
        time.sleep(2.0)

  def postMessage(self, message):
    req = urllib2.Request(self.url)
    req.add_header('Content-Type', 'application/json')
    try:
      urllib2.urlopen(req, json.dumps(message))
    except urllib2.URLError as e:
      print e

File: src/ruby/messenger.rb
require 'net/http'
require 'json'

class Messenger
  def initialize(url)
    @url = URI(url)
  end

  def getMessage
    while true do

      begin
        response = Net::HTTP.get_response(@url)
        if response.code == '200' then
          return JSON.parse(response.body)
        end
      rescue JSON::ParserError
        sleep 2
      rescue Errno::ECONNREFUSED
        sleep 2
      rescue EOFError
        sleep 2
      end

    end
  end

  def sendMessage(message)
    print "Sending '" + message.to_s + "\n"
    begin
      http = Net::HTTP.new(@url.host, @url.port)
      req = Net::HTTP::Post.new(@url.path, initheader = {'Content-type' => 'application/json'})
      req.body = message.to_json
      http.request(req)
    rescue Errno::ECONNREFUSED
    rescue EOFError
    end
  end
end
